# 알츠하이머 예측하기 
## 데이터 출처
- 알츠하이머 데이터
- https://www.kaggle.com/datasets/rabieelkharoua/alzheimers-disease-dataset

# 데이터 확인
## 1. 인적 정보
|column 명|설명|value|column 명|설명|value|
|--|--|--|--|--|--|
|PatientID|식별 번호|471-6900|Age|나이|60-90|
|Gender|성별|0:남자, 1 : 여자|Ethnicity|인종|0:백인, 1 : 아프리카계 미국인, 2 : 아시아, 3 : 기타|
|EducationLevel|교육 수준|0 : 없음, 1 : 고등학교, 2 : 학사, 3 : 더 높음|

## 2. lifestype
|column 명|설명|value|column 명|설명|value|
|--|--|--|--|--|--|
|Smoking|흡연 유/무|0 : 아니오, 1 : 예|AlcoholConsumption|주당 알코올 소비량|0-20|
|PhysicalActivity|주간 신체 활동량|0-10|DietQuality|식단 품질|0-10|
|SleepQuality|수면의 질|4-10|

## 3. 병력
|column 명|설명|value|column 명|설명|value|
|--|--|--|--|--|--|
|FamilyHistoryAlzheimers|알츠하이머 가족력|0 : 없음, 1 : 예|CardiovascularDisease|심혈관 질환 유무|0 : 없음, 1 : 예|
|Diabetes|당뇨병 유무|0 : 없음, 1 : 예|Depression|우울증 유무|0 : 없음, 1 : 예|
|HeadInjury|두부 손상 유무|0 : 없음, 1 : 예|Hypertension|고혈압 유무|0 : 없음, 1 : 예|

## 4. 임상 측정
|column 명|설명|value|column 명|설명|value|
|--|--|--|--|--|--|
|SystolicBP|수축기 혈압|90-180|DiastolicBP|이완기 혈압|60-120|
|CholesterolTotal|콜레스테롤 총량|150-300|CholesterolLDL|LDL콜레스테롤 수치|50-200|
|CholesterolHDL|HDL콜레스테롤 수치|20-100|CholesterolTriglycerides|중성지방|50-400|
|BMI|신체질량지수|15-40|

## 5. 인지 및 기능 평가
|column 명|설명|value|column 명|설명|value|
|--|--|--|--|--|--|
|MMSE|간이 정신 상태 검사 점수|0-30, 낮을수록 인지 장애가 있음|FunctionalAssessment|기능평가|0-10, 낮을수록 기능 장애가 심함|
|MemoryComplaints|기억 장애 유무|0 : 없음, 1 : 예|BehavioralProblems|행동 문제 유무|0 : 없음, 1 : 예|
|ADL|일상생활활동 점수|0-10, 낮을수록 장애가 심함|

## 6. 증상
|column 명|설명|value|column 명|설명|value|
|--|--|--|--|--|--|
|Confusion|혼란의 유무|0 : 없음, 1 : 예|Disorientation|방향 감각 상실 유무|0 : 없음, 1 : 예|
|PersonalityChanges|성격 변화 유무|0 : 없음, 1 : 예|DifficultyCompletingTasks|작업 완료 어려움 유무|0 : 없음, 1 : 예|
|Forgetfulness|건망증 유무|0 : 없음, 1 : 예|

## 7. 진단
|column 명|설명|value|column 명|설명|value|
|--|--|--|--|--|--|
|Diagnosis|알츠하이머 유무|0 : 없음, 1 : 예|DoctorInCharge|담당 의사에 대한 정보|'XXXConfid'|

In [1]:
pip install statsmodels


[notice] A new release of pip is available: 24.0 -> 24.1.1
[notice] To update, run: pip install --upgrade pip
Note: you may need to restart the kernel to use updated packages.


In [2]:
pip install imblearn


[notice] A new release of pip is available: 24.0 -> 24.1.1
[notice] To update, run: pip install --upgrade pip
Note: you may need to restart the kernel to use updated packages.


In [3]:
import pandas as pd
import numpy as np
from scipy import stats 
from statsmodels.formula.api import logit 
from sklearn.metrics import accuracy_score,classification_report
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split


df = pd.read_csv('../data/alzheimers_disease_data.csv')
df

PatientID  Age  Gender  Ethnicity  EducationLevel        BMI  Smoking  \
0          4751   73       0          0               2  22.927749        0   
1          4752   89       0          0               0  26.827681        0   
2          4753   73       0          3               1  17.795882        0   
3          4754   74       1          0               1  33.800817        1   
4          4755   89       0          0               0  20.716974        0   
...         ...  ...     ...        ...             ...        ...      ...   
2144       6895   61       0          0               1  39.121757        0   
2145       6896   75       0          0               2  17.857903        0   
2146       6897   77       0          0               1  15.476479        0   
2147       6898   78       1          3               1  15.299911        0   
2148       6899   72       0          0               2  33.289738        0   

      AlcoholConsumption  PhysicalActivity  DietQuality  ...  \
0              13.297218          6.327112     1.347214  ...   
1               4.542524          7.619885     0.518767  ...   
2              19.555085          7.844988     1.826335  ...   
3              12.209266          8.428001     7.435604  ...   
4              18.454356          6.310461     0.795498  ...   
...                  ...               ...          ...  ...   
2144            1.561126          4.049964     6.555306  ...   
2145           18.767261          1.360667     2.904662  ...   
2146            4.594670          9.886002     8.120025  ...   
2147            8.674505          6.354282     1.263427  ...   
2148            7.890703          6.570993     7.941404  ...   

      MemoryComplaints  BehavioralProblems       ADL  Confusion  \
0                    0                   0  1.725883          0   
1                    0                   0  2.592424          0   
2                    0                   0  7.119548          0   
3                    0                   1  6.481226          0   
4                    0                   0  0.014691          0   
...                ...                 ...       ...        ...   
2144                 0                   0  4.492838          1   
2145                 0                   1  9.204952          0   
2146                 0                   0  5.036334          0   
2147                 0                   0  3.785399          0   
2148                 0                   1  8.327563          0   

      Disorientation  PersonalityChanges  DifficultyCompletingTasks  \
0                  0                   0                          1   
1                  0                   0                          0   
2                  1                   0                          1   
3                  0                   0                          0   
4                  0                   1                          1   
...              ...                 ...                        ...   
2144               0                   0                          0   
2145               0                   0                          0   
2146               0                   0                          0   
2147               0                   0                          0   
2148               1                   0                          0   

      Forgetfulness  Diagnosis  DoctorInCharge  
0                 0          0       XXXConfid  
1                 1          0       XXXConfid  
2                 0          0       XXXConfid  
3                 0          0       XXXConfid  
4                 0          0       XXXConfid  
...             ...        ...             ...  
2144              0          1       XXXConfid  
2145              0          1       XXXConfid  
2146              0          1       XXXConfid  
2147              1          1       XXXConfid  
2148              1          0       XXXConfid  

[2149 rows x 35 columns]

In [4]:
df.describe().T

count         mean         std          min  \
PatientID                  2149.0  5825.000000  620.507185  4751.000000   
Age                        2149.0    74.908795    8.990221    60.000000   
Gender                     2149.0     0.506282    0.500077     0.000000   
Ethnicity                  2149.0     0.697534    0.996128     0.000000   
EducationLevel             2149.0     1.286645    0.904527     0.000000   
BMI                        2149.0    27.655697    7.217438    15.008851   
Smoking                    2149.0     0.288506    0.453173     0.000000   
AlcoholConsumption         2149.0    10.039442    5.757910     0.002003   
PhysicalActivity           2149.0     4.920202    2.857191     0.003616   
DietQuality                2149.0     4.993138    2.909055     0.009385   
SleepQuality               2149.0     7.051081    1.763573     4.002629   
FamilyHistoryAlzheimers    2149.0     0.252210    0.434382     0.000000   
CardiovascularDisease      2149.0     0.144253    0.351428     0.000000   
Diabetes                   2149.0     0.150768    0.357906     0.000000   
Depression                 2149.0     0.200558    0.400511     0.000000   
HeadInjury                 2149.0     0.092601    0.289940     0.000000   
Hypertension               2149.0     0.148906    0.356079     0.000000   
SystolicBP                 2149.0   134.264774   25.949352    90.000000   
DiastolicBP                2149.0    89.847836   17.592496    60.000000   
CholesterolTotal           2149.0   225.197519   42.542233   150.093316   
CholesterolLDL             2149.0   124.335944   43.366584    50.230707   
CholesterolHDL             2149.0    59.463533   23.139174    20.003434   
CholesterolTriglycerides   2149.0   228.281496  101.986721    50.407194   
MMSE                       2149.0    14.755132    8.613151     0.005312   
FunctionalAssessment       2149.0     5.080055    2.892743     0.000460   
MemoryComplaints           2149.0     0.208004    0.405974     0.000000   
BehavioralProblems         2149.0     0.156817    0.363713     0.000000   
ADL                        2149.0     4.982958    2.949775     0.001288   
Confusion                  2149.0     0.205212    0.403950     0.000000   
Disorientation             2149.0     0.158213    0.365026     0.000000   
PersonalityChanges         2149.0     0.150768    0.357906     0.000000   
DifficultyCompletingTasks  2149.0     0.158678    0.365461     0.000000   
Forgetfulness              2149.0     0.301536    0.459032     0.000000   
Diagnosis                  2149.0     0.353653    0.478214     0.000000   

                                   25%          50%          75%          max  
PatientID                  5288.000000  5825.000000  6362.000000  6899.000000  
Age                          67.000000    75.000000    83.000000    90.000000  
Gender                        0.000000     1.000000     1.000000     1.000000  
Ethnicity                     0.000000     0.000000     1.000000     3.000000  
EducationLevel                1.000000     1.000000     2.000000     3.000000  
BMI                          21.611408    27.823924    33.869778    39.992767  
Smoking                       0.000000     0.000000     1.000000     1.000000  
AlcoholConsumption            5.139810     9.934412    15.157931    19.989293  
PhysicalActivity              2.570626     4.766424     7.427899     9.987429  
DietQuality                   2.458455     5.076087     7.558625     9.998346  
SleepQuality                  5.482997     7.115646     8.562521     9.999840  
FamilyHistoryAlzheimers       0.000000     0.000000     1.000000     1.000000  
CardiovascularDisease         0.000000     0.000000     0.000000     1.000000  
Diabetes                      0.000000     0.000000     0.000000     1.000000  
Depression                    0.000000     0.000000     0.000000     1.000000  
HeadInjury                    0.000000     0.000000     0.000000     1.000000  
Hypertension                  0.000000    

In [5]:
df.EducationLevel.value_counts()

EducationLevel
1    854
2    636
0    446
3    213
Name: count, dtype: int64

# 데이터 전처리

## 결측치 확인 : 없음

In [6]:
df.isnull().sum()

PatientID                    0
Age                          0
Gender                       0
Ethnicity                    0
EducationLevel               0
BMI                          0
Smoking                      0
AlcoholConsumption           0
PhysicalActivity             0
DietQuality                  0
SleepQuality                 0
FamilyHistoryAlzheimers      0
CardiovascularDisease        0
Diabetes                     0
Depression                   0
HeadInjury                   0
Hypertension                 0
SystolicBP                   0
DiastolicBP                  0
CholesterolTotal             0
CholesterolLDL               0
CholesterolHDL               0
CholesterolTriglycerides     0
MMSE                         0
FunctionalAssessment         0
MemoryComplaints             0
BehavioralProblems           0
ADL                          0
Confusion                    0
Disorientation               0
PersonalityChanges           0
DifficultyCompletingTasks    0
Forgetfu

## 필요없는 column 제거 : PatientID, DoctorInCharge

In [7]:
df.drop(['PatientID','DoctorInCharge'],axis=1, inplace=True)

## 데이터 정규화

In [8]:
scaler = MinMaxScaler()
for i in ['Age','BMI','AlcoholConsumption','PhysicalActivity','DietQuality','SleepQuality','SystolicBP','DiastolicBP','DiastolicBP','CholesterolTotal','CholesterolLDL','CholesterolHDL','CholesterolTriglycerides','MMSE','FunctionalAssessment','ADL']:
    df[i] = scaler.fit_transform(df[[i]])
df.describe()

Age       Gender    Ethnicity  EducationLevel          BMI  \
count  2149.000000  2149.000000  2149.000000     2149.000000  2149.000000   
mean      0.496960     0.506282     0.697534        1.286645     0.506199   
std       0.299674     0.500077     0.996128        0.904527     0.288883   
min       0.000000     0.000000     0.000000        0.000000     0.000000   
25%       0.233333     0.000000     0.000000        1.000000     0.264272   
50%       0.500000     1.000000     0.000000        1.000000     0.512933   
75%       0.766667     1.000000     1.000000        2.000000     0.754923   
max       1.000000     1.000000     3.000000        3.000000     1.000000   

           Smoking  AlcoholConsumption  PhysicalActivity  DietQuality  \
count  2149.000000         2149.000000       2149.000000  2149.000000   
mean      0.288506            0.502191          0.492456     0.498926   
std       0.453173            0.288079          0.286182     0.291227   
min       0.000000            0.000000          0.000000     0.000000   
25%       0.000000            0.257054          0.257117     0.245178   
50%       0.000000            0.496936          0.477053     0.507230   
75%       1.000000            0.758278          0.743632     0.755758   
max       1.000000            1.000000          1.000000     1.000000   

       SleepQuality  ...  FunctionalAssessment  MemoryComplaints  \
count   2149.000000  ...           2149.000000       2149.000000   
mean       0.508312  ...              0.508162          0.208004   
std        0.294065  ...              0.289390          0.405974   
min        0.000000  ...              0.000000          0.000000   
25%        0.246843  ...              0.256685          0.000000   
50%        0.519077  ...              0.509601          0.000000   
75%        0.760335  ...              0.754954          0.000000   
max        1.000000  ...              1.000000          1.000000   

       BehavioralProblems          ADL    Confusion  Disorientation  \
count         2149.000000  2149.000000  2149.000000     2149.000000   
mean             0.156817     0.498244     0.205212        0.158213   
std              0.363713     0.295023     0.403950        0.365026   
min              0.000000     0.000000     0.000000        0.000000   
25%              0.000000     0.234191     0.000000        0.000000   
50%              0.000000     0.503846     0.000000        0.000000   
75%              0.000000     0.758137     0.000000        0.000000   
max              1.000000     1.000000     1.000000        1.000000   

       PersonalityChanges  DifficultyCompletingTasks  Forgetfulness  \
count         2149.000000                2149.000000    2149.000000   
mean             0.150768                   0.158678       0.301536   
std              0.357906                   0.365461       0.459032   
min              0.000000                   0.000000       0.000000   
25%              0.000000                   0.000000       0.000000   
50%              0.000000                   0.000000       0.000000   
75%              0.000000                   0.000000       1.000000   
max              1.000000                   1.000000       1.000000   

         Diagnosis  
count  2149.000000  
mean      0.353653  
std       0.478214  
min       0.000000  
25%       0.000000  
50%       0.000000  
75%       1.000000  
max       1.000000  

[8 rows x 33 columns]

## 데이터 샘플링 : SMOTE

In [9]:
binary_columns = ['Gender','Ethnicity','EducationLevel','Smoking','FamilyHistoryAlzheimers','CardiovascularDisease','Diabetes','Depression','HeadInjury','Hypertension','MemoryComplaints','BehavioralProblems','Confusion','Disorientation','PersonalityChanges','DifficultyCompletingTasks','Forgetfulness']
binary_indices = [df.columns.get_loc(col) for col in binary_columns]
binary_indices

[1, 2, 3, 5, 10, 11, 12, 13, 14, 15, 24, 25, 27, 28, 29, 30, 31]

In [10]:
from imblearn.over_sampling import SMOTENC
import pandas as pd


# SMOTENC 적용
overSampling = SMOTENC(categorical_features=binary_indices, sampling_strategy=0.8, random_state=42)
feature = df.drop('Diagnosis', axis=1)
target = df['Diagnosis']
feature_sample, target_sample = overSampling.fit_resample(feature, target)

print(feature.shape, target.shape, feature_sample.shape, target_sample.shape)
df = pd.concat([feature_sample, target_sample], axis=1)
df


(2149, 32) (2149,) (2500, 32) (2500,)


Age  Gender  Ethnicity  EducationLevel       BMI  Smoking  \
0     0.433333       0          0               2  0.316960        0   
1     0.966667       0          0               0  0.473058        0   
2     0.433333       0          3               1  0.111553        0   
3     0.466667       1          0               1  0.752163        1   
4     0.966667       0          0               0  0.228472        0   
...        ...     ...        ...             ...       ...      ...   
2495  0.105265       1          0               2  0.662061        0   
2496  0.522259       1          0               1  0.841854        0   
2497  0.151783       1          0               0  0.167029        0   
2498  0.552156       1          0               2  0.782612        0   
2499  0.733526       0          0               2  0.819932        0   

      AlcoholConsumption  PhysicalActivity  DietQuality  SleepQuality  ...  \
0               0.665183          0.633375     0.133931      0.837564  ...   
1               0.227170          0.762862     0.050995      0.525021  ...   
2               0.978276          0.785408     0.181896      0.945597  ...   
3               0.610751          0.843804     0.743443      0.731994  ...   
4               0.923204          0.631707     0.078698      0.265892  ...   
...                  ...               ...          ...           ...  ...   
2495            0.060032          0.503047     0.287995      0.162242  ...   
2496            0.828460          0.310607     0.905753      0.642736  ...   
2497            0.734901          0.278921     0.350008      0.670189  ...   
2498            0.837168          0.613912     0.624380      0.754645  ...   
2499            0.384893          0.063661     0.761883      0.684130  ...   

      FunctionalAssessment  MemoryComplaints  BehavioralProblems       ADL  \
0                 0.652102                 0                   0  0.172486   
1                 0.712108                 0                   0  0.259154   
2                 0.589697                 0                   0  0.711936   
3                 0.896823                 0                   1  0.648094   
4                 0.604699                 0                   0  0.001341   
...                    ...               ...                 ...       ...   
2495              0.290674                 0                   0  0.164272   
2496              0.192537                 0                   0  0.098954   
2497              0.737102                 0                   0  0.553522   
2498              0.709766                 1                   0  0.106809   
2499              0.176169                 0                   0  0.481622   

      Confusion  Disorientation  PersonalityChanges  \
0             0               0                   0   
1             0               0                   0   
2             0               1                   0   
3             0               0                   0   
4             0               0                   1   
...         ...             ...                 ...   
2495          0               0                   0   
2496          0               0                   0   
2497          0               0                   0   
2498          0               0                   0   
2499          0               0                   0   

      DifficultyCompletingTasks  Forgetfulness  Diagnosis  
0                             1              0          0  
1                             0              1          0  
2                             1              0          0  
3                             0              0          0  
4                             1              0          0  
...                         ...            ...        ...  
2495                          0              0          1  
2496                          0              0          1  
2497                          0              0          1  
2498       

# 가설 설정 

## 1. 증상이 알츠하이머의 유/무에 영향을 미친다. 
- 방법 : 증상으로 알츠하이머의 유/무를 예측하는 로지스틱 회귀 모형을 만들어 각 독립변수들의 p-value를 통해 종속 변수에 영향을 미치는지 확인
- 성능 지표 : 로지스틱 회귀 모형의 정확성이 0.65로 높은 성능을 보이지 않음
- 유의미한 변수 : 불안, 방향 감각 상실, 성격 변화, 건망증
    - 불안 : -0.390
    - 방향 감각 상실 : -0.475
    - 성격 변화 : - 0.540
    - 건망증 : -0.296

In [11]:
from statsmodels.formula.api import logit 
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
df_symptom = df[['Confusion','Disorientation','PersonalityChanges','DifficultyCompletingTasks','Forgetfulness','Diagnosis']]
train,test = train_test_split(df_symptom,test_size = 0.3, random_state = 23)
model = logit('Diagnosis ~ Confusion + Disorientation + PersonalityChanges + DifficultyCompletingTasks + Forgetfulness',train).fit()
pred = model.predict(test)
pred_class = (pred > 0.5).astype(int)
print(classification_report(test['Diagnosis'],pred_class))

Optimization terminated successfully.
         Current function value: 0.672219
         Iterations 5
              precision    recall  f1-score   support

           0       0.61      0.69      0.65       403
           1       0.57      0.48      0.52       347

    accuracy                           0.59       750
   macro avg       0.59      0.59      0.59       750
weighted avg       0.59      0.59      0.59       750



### 불안

In [12]:

if model.pvalues['Confusion'] >= 0.05:
    print('p-value : ',model.pvalues['Confusion'])
    print('귀무가설, 종속 변수에 영향을 미치지 않음')
else:
    print('회귀 계수 : ',model.params['Confusion'])
    print('p-value : ',model.pvalues['Confusion'])
    print('대립가설, 종속 변수에 영향을 미침')

회귀 계수 :  -0.3902876540906309
p-value :  0.0026550551939513068
대립가설, 종속 변수에 영향을 미침


### 방향 감각 상실

In [13]:
if model.pvalues['Disorientation'] >= 0.05:
    print('p-value : ',model.pvalues['Disorientation'])
    print('귀무가설, 종속 변수에 영향을 미치지 않음')
else:
    print('회귀 계수 : ',model.params['Disorientation'])
    print('p-value : ',model.pvalues['Disorientation'])
    print('대립가설, 종속 변수에 영향을 미침')

회귀 계수 :  -0.4750624193945828
p-value :  0.0012722456763638197
대립가설, 종속 변수에 영향을 미침


### 성격 변화

In [14]:
if model.pvalues['PersonalityChanges'] >= 0.05:
    print('p-value : ',model.pvalues['PersonalityChanges'])
    print('귀무가설, 종속 변수에 영향을 미치지 않음')
else:
    print('회귀 계수 : ',model.params['PersonalityChanges'])
    print('p-value : ',model.pvalues['PersonalityChanges'])
    print('대립가설, 종속 변수에 영향을 미침')

회귀 계수 :  -0.54019998169794
p-value :  0.00045615052330734684
대립가설, 종속 변수에 영향을 미침


### 작업 완료 어려움

In [15]:
if model.pvalues['DifficultyCompletingTasks'] >= 0.05:
    print('p-value : ',model.pvalues['DifficultyCompletingTasks'])
    print('귀무가설, 종속 변수에 영향을 미치지 않음')
else:
    print('회귀 계수 : ',model.params['DifficultyCompletingTasks'])
    print('p-value : ',model.pvalues['DifficultyCompletingTasks'])
    print('대립가설, 종속 변수에 영향을 미침')

p-value :  0.10775778530120357
귀무가설, 종속 변수에 영향을 미치지 않음


### 건망증

In [16]:
if model.pvalues['Forgetfulness'] >= 0.05:
    print('p-value : ',model.pvalues['Forgetfulness'])
    print('귀무가설, 종속 변수에 영향을 미치지 않음')
else:
    print('회귀 계수 : ',model.params['Forgetfulness'])
    print('p-value : ',model.pvalues['Forgetfulness'])
    print('대립가설, 종속 변수에 영향을 미침')

회귀 계수 :  -0.295734419166147
p-value :  0.006924161148549263
대립가설, 종속 변수에 영향을 미침


## 2. 인지 및 기능 평가 변수들이 알츠하이머의 유/무에 영향을 미친다.
- 방법 : 인지 및 기능 평가 변수들로 알츠하이머의 유/무를 예측하는 로지스틱 회귀 모형을 만들어 각 독립변수들의 p-value를 통해 종속 변수에 영향을 미치는지 확인
- 성능 지표 : 로지스틱 회귀 모형의 정확성이 0.84로 높은 성능을 보임
- 유의미한 변수 : 간이 점수 상태 점수, 기능 평가, 기억 장애 유무, 행동 문제 유무, 일상생활 활동 점수
    - 간이 점수 상태 점수 : -2.936
    - 기능 평가 : -4.150
    - 기억 장애 유무 : 2.199
    - 행동 문제 유무 : 1.929
    - 일상생활 활동 점수 : -3.979

In [17]:
from statsmodels.formula.api import logit 
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
df_assessment = df[['MMSE','FunctionalAssessment','MemoryComplaints','BehavioralProblems','ADL','Diagnosis']]
train,test = train_test_split(df_assessment,test_size = 0.3, random_state = 23)
model = logit('Diagnosis ~ MMSE + FunctionalAssessment + MemoryComplaints + BehavioralProblems + ADL',train).fit()
pred = model.predict(test)
pred_class = (pred > 0.5).astype(int)
print(classification_report(test['Diagnosis'],pred_class))

Optimization terminated successfully.
         Current function value: 0.430391
         Iterations 7
              precision    recall  f1-score   support

           0       0.85      0.86      0.85       403
           1       0.83      0.82      0.83       347

    accuracy                           0.84       750
   macro avg       0.84      0.84      0.84       750
weighted avg       0.84      0.84      0.84       750



### 간이 점수 상태 점수

In [18]:
if model.pvalues['MMSE'] >= 0.05:
    print('p-value : ',model.pvalues['MMSE'])
    print('귀무가설, 종속 변수에 영향을 미치지 않음')
else:
    print('회귀 계수 : ',model.params['MMSE'])
    print('p-value : ',model.pvalues['MMSE'])
    print('대립가설, 종속 변수에 영향을 미침')

회귀 계수 :  -2.9356309285588638
p-value :  2.351967229612939e-31
대립가설, 종속 변수에 영향을 미침


### 기능 평가

In [19]:

if model.pvalues['FunctionalAssessment'] >= 0.05:
    print('p-value : ',model.pvalues['FunctionalAssessment'])
    print('귀무가설, 종속 변수에 영향을 미치지 않음')
else:
    print('회귀 계수 : ',model.params['FunctionalAssessment'])
    print('p-value : ',model.pvalues['FunctionalAssessment'])
    print('대립가설, 종속 변수에 영향을 미침')

회귀 계수 :  -4.149696237514159
p-value :  3.1273086107795373e-55
대립가설, 종속 변수에 영향을 미침


### 기억 장애 유무

In [20]:
if model.pvalues['MemoryComplaints'] >= 0.05:
    print('p-value : ',model.pvalues['MemoryComplaints'])
    print('귀무가설, 종속 변수에 영향을 미치지 않음')
else:
    print('회귀 계수 : ',model.params['MemoryComplaints'])
    print('p-value : ',model.pvalues['MemoryComplaints'])
    print('대립가설, 종속 변수에 영향을 미침')

회귀 계수 :  2.1991199114630846
p-value :  2.90883796310243e-38
대립가설, 종속 변수에 영향을 미침


### 행동 문제 유무

In [21]:
if model.pvalues['BehavioralProblems'] >= 0.05:
    print('p-value : ',model.pvalues['BehavioralProblems'])
    print('귀무가설, 종속 변수에 영향을 미치지 않음')
else:
    print('회귀 계수 : ',model.params['BehavioralProblems'])
    print('p-value : ',model.pvalues['BehavioralProblems'])
    print('대립가설, 종속 변수에 영향을 미침')

회귀 계수 :  1.9289703744234608
p-value :  3.8487657013726707e-23
대립가설, 종속 변수에 영향을 미침


### 일상생활활동 점수

In [22]:

if model.pvalues['ADL'] >= 0.05:
    print('p-value : ',model.pvalues['ADL'])
    print('귀무가설, 종속 변수에 영향을 미치지 않음')
else:
    print('회귀 계수 : ',model.params['ADL'])
    print('p-value : ',model.pvalues['ADL'])    
    print('대립가설, 종속 변수에 영향을 미침')

회귀 계수 :  -3.9793562730596563
p-value :  1.5417643942503525e-52
대립가설, 종속 변수에 영향을 미침


## 3. 임상 측정 결과 변수들이 알츠하이머의 유/무에 영향을 미친다.
- 방법 : 임상 측정 결과 변수들로 알츠하이머의 유/무를 예측하는 로지스틱 회귀 모형을 만들어 각 독립변수들의 p-value를 통해 종속 변수에 영향을 미치는지 확인
- 성능 지표 : 로지스틱 회귀 모형의 정확성이 0.53로 높은 성능을 보이지 않음,f1-score 역시 상당히 낮음
- 유의미한 변수 : 없음

In [23]:
from statsmodels.formula.api import logit 
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
df_clinical = df[['SystolicBP','DiastolicBP','CholesterolTotal','CholesterolLDL','CholesterolHDL','CholesterolTriglycerides','BMI','Diagnosis']]
train,test = train_test_split(df_clinical,test_size = 0.3, random_state = 23)
model = logit('Diagnosis ~ SystolicBP + DiastolicBP + CholesterolTotal + CholesterolLDL + CholesterolHDL + CholesterolTriglycerides + BMI',train).fit()
pred = model.predict(test)
pred_class = (pred > 0.5).astype(int)
print(classification_report(test['Diagnosis'],pred_class))

Optimization terminated successfully.
         Current function value: 0.682324
         Iterations 4
              precision    recall  f1-score   support

           0       0.54      0.95      0.69       403
           1       0.44      0.04      0.08       347

    accuracy                           0.53       750
   macro avg       0.49      0.50      0.38       750
weighted avg       0.49      0.53      0.41       750



### 수축기 혈압

In [24]:
if model.pvalues['SystolicBP'] >= 0.05:
    print('p-value : ',model.pvalues['SystolicBP'])
    print('귀무가설, 종속 변수에 영향을 미치지 않음')
else:
    print('회귀 계수 : ',model.params['SystolicBP'])
    print('p-value : ',model.pvalues['SystolicBP'])
    print('대립가설, 종속 변수에 영향을 미침')

p-value :  0.6940000678705067
귀무가설, 종속 변수에 영향을 미치지 않음


### 이완기 혈압

In [25]:
if model.pvalues['DiastolicBP'] >= 0.05:
    print('p-value : ',model.pvalues['DiastolicBP'])
    print('귀무가설, 종속 변수에 영향을 미치지 않음')
else:
    print('회귀 계수 : ',model.params['DiastolicBP'])
    print('p-value : ',model.pvalues['DiastolicBP'])
    print('대립가설, 종속 변수에 영향을 미침')

p-value :  0.5385014178286069
귀무가설, 종속 변수에 영향을 미치지 않음


### 콜레스테롤 총량

In [26]:
if model.pvalues['CholesterolTotal'] >= 0.05:
    print('p-value : ',model.pvalues['CholesterolTotal'])
    print('귀무가설, 종속 변수에 영향을 미치지 않음')
else:
    print('회귀 계수 : ',model.params['CholesterolTotal'])
    print('p-value : ',model.pvalues['CholesterolTotal'])
    print('대립가설, 종속 변수에 영향을 미침')

p-value :  0.19624603565036647
귀무가설, 종속 변수에 영향을 미치지 않음


### 콜레스테롤 LDL

In [27]:
if model.pvalues['CholesterolLDL'] >= 0.05:
    print('p-value : ',model.pvalues['CholesterolLDL'])
    print('귀무가설, 종속 변수에 영향을 미치지 않음')
else:
    print('회귀 계수 : ',model.params['CholesterolLDL'])
    print('p-value : ',model.pvalues['CholesterolLDL'])
    print('대립가설, 종속 변수에 영향을 미침')

p-value :  0.25979717470493924
귀무가설, 종속 변수에 영향을 미치지 않음


### 콜레스테롤 HDL

In [28]:
if model.pvalues['CholesterolHDL'] >= 0.05:
    print('p-value : ',model.pvalues['CholesterolHDL'])
    print('귀무가설, 종속 변수에 영향을 미치지 않음')
else:
    print('회귀 계수 : ',model.params['CholesterolHDL'])
    print('p-value : ',model.pvalues['CholesterolHDL'])
    print('대립가설, 종속 변수에 영향을 미침')

p-value :  0.10383465718795826
귀무가설, 종속 변수에 영향을 미치지 않음


### 중성 지방

In [29]:
if model.pvalues['CholesterolTriglycerides'] >= 0.05:
    print('p-value : ',model.pvalues['CholesterolTriglycerides'])
    print('귀무가설, 종속 변수에 영향을 미치지 않음')
else:
    print('회귀 계수 : ',model.params['CholesterolTriglycerides'])
    print('p-value : ',model.pvalues['CholesterolTriglycerides'])
    print('대립가설, 종속 변수에 영향을 미침')

p-value :  0.5646059755177482
귀무가설, 종속 변수에 영향을 미치지 않음


### BMI

In [30]:
if model.pvalues['BMI'] >= 0.05:
    print('p-value : ',model.pvalues['BMI'])
    print('귀무가설, 종속 변수에 영향을 미치지 않음')
else:
    print('회귀 계수 : ',model.params['BMI'])
    print('p-value : ',model.pvalues['BMI'])
    print('대립가설, 종속 변수에 영향을 미침')

p-value :  0.09020795071721101
귀무가설, 종속 변수에 영향을 미치지 않음


## 4. 병력 변수들이 알츠하이머의 유/무에 영향을 미친다.
- 방법 : 임상 측정 결과 변수들로 알츠하이머의 유/무를 예측하는 로지스틱 회귀 모형을 만들어 각 독립변수들의 p-value를 통해 종속 변수에 영향을 미치는지 확인
- 성능 지표 : 로지스틱 회귀 모형의 정확성이 0.58로 높은 성능을 보이지 않음
- 유의미한 변수 : 알츠하이머 가족력, 당뇨병, 우울증, 두부 손상
    - 알츠하이머 가족력 : -0.509
    - 당뇨병 : -0.534
    - 우울증 : -0.346
    - 두부 손상 : -0.662

In [31]:
from statsmodels.formula.api import logit 
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
df_history = df[['FamilyHistoryAlzheimers','CardiovascularDisease','Diabetes','Depression','HeadInjury','Hypertension','Diagnosis']]
train,test = train_test_split(df_history,test_size = 0.3, random_state = 23)
model = logit('Diagnosis ~ FamilyHistoryAlzheimers + CardiovascularDisease + Diabetes + Depression + HeadInjury + Hypertension',train).fit()
pred = model.predict(test)
pred_class = (pred > 0.5).astype(int)
print(classification_report(test['Diagnosis'],pred_class))

Optimization terminated successfully.
         Current function value: 0.669145
         Iterations 5
              precision    recall  f1-score   support

           0       0.59      0.68      0.63       403
           1       0.55      0.46      0.50       347

    accuracy                           0.58       750
   macro avg       0.57      0.57      0.57       750
weighted avg       0.57      0.58      0.57       750



### 알츠하이머 가족력

In [32]:
if model.pvalues['FamilyHistoryAlzheimers'] >= 0.05:
    print('p-value : ',model.pvalues['FamilyHistoryAlzheimers'])
    print('귀무가설, 종속 변수에 영향을 미치지 않음')
else:
    print('회귀 계수 : ',model.params['FamilyHistoryAlzheimers'])
    print('p-value : ',model.pvalues['FamilyHistoryAlzheimers'])
    print('대립가설, 종속 변수에 영향을 미침')

회귀 계수 :  -0.5091993159957983
p-value :  2.5795046599213816e-05
대립가설, 종속 변수에 영향을 미침


### 심혈관 질환

In [33]:
if model.pvalues['CardiovascularDisease'] >= 0.05:
    print('p-value : ',model.pvalues['CardiovascularDisease'])
    print('귀무가설, 종속 변수에 영향을 미치지 않음')
else:
    print('회귀 계수 : ',model.params['CardiovascularDisease'])
    print('p-value : ',model.pvalues['CardiovascularDisease'])
    print('대립가설, 종속 변수에 영향을 미침')

p-value :  0.3756456001655083
귀무가설, 종속 변수에 영향을 미치지 않음


### 당뇨병

In [34]:
if model.pvalues['Diabetes'] >= 0.05:
    print('p-value : ',model.pvalues['Diabetes'])
    print('귀무가설, 종속 변수에 영향을 미치지 않음')
else:
    print('회귀 계수 : ',model.params['Diabetes'])
    print('p-value : ',model.pvalues['Diabetes'])
    print('대립가설, 종속 변수에 영향을 미침')

회귀 계수 :  -0.5336545557021806
p-value :  0.0006108416490310457
대립가설, 종속 변수에 영향을 미침


### 우울증

In [35]:
if model.pvalues['Depression'] >= 0.05:
    print('p-value : ',model.pvalues['Depression'])
    print('귀무가설, 종속 변수에 영향을 미치지 않음')
else:
    print('회귀 계수 : ',model.params['Depression'])
    print('p-value : ',model.pvalues['Depression'])
    print('대립가설, 종속 변수에 영향을 미침')

회귀 계수 :  -0.34603502344657366
p-value :  0.007461414118246072
대립가설, 종속 변수에 영향을 미침


### 두부 손상

In [36]:
if model.pvalues['HeadInjury'] >= 0.05:
    print('p-value : ',model.pvalues['HeadInjury'])
    print('귀무가설, 종속 변수에 영향을 미치지 않음')
else:
    print('회귀 계수 : ',model.params['HeadInjury'])
    print('p-value : ',model.pvalues['HeadInjury'])
    print('대립가설, 종속 변수에 영향을 미침')

회귀 계수 :  -0.6618080553366018
p-value :  0.0008779593213629934
대립가설, 종속 변수에 영향을 미침


### 고혈압

In [37]:
if model.pvalues['Hypertension'] >= 0.05:
    print('p-value : ',model.pvalues['Hypertension'])
    print('귀무가설, 종속 변수에 영향을 미치지 않음')
else:
    print('회귀 계수 : ',model.params['Hypertension'])
    print('p-value : ',model.pvalues['Hypertension'])
    print('대립가설, 종속 변수에 영향을 미침')

p-value :  0.14340262428827716
귀무가설, 종속 변수에 영향을 미치지 않음


## 5. 일상생활 변수들이 알츠하이머의 유/무에 영향을 미친다.
- 방법 : 일상생활 변수들로 알츠하이머의 유/무를 예측하는 로지스틱 회귀 모형을 만들어 각 독립변수들의 p-value를 통해 종속 변수에 영향을 미치는지 확인
- 성능 지표 : 로지스틱 회귀 모형의 정확성이 0.54로 높은 성능을 보이지 않음, f1 score 역시 상당히 낮음
- 유의미한 변수 :흡여 유/무
    - 흡연 유/무 : -0.337

In [50]:
from statsmodels.formula.api import logit 
from sklearn.metrics import accuracy_score,f1_score
from sklearn.model_selection import train_test_split
df_lifestyle = df[['Smoking','AlcoholConsumption','PhysicalActivity','DietQuality','SleepQuality','Diagnosis']]
train,test = train_test_split(df_lifestyle,test_size = 0.3, random_state = 23)
model = logit('Diagnosis ~ Smoking + AlcoholConsumption + PhysicalActivity + DietQuality + SleepQuality',train).fit()
pred = model.predict(test)
pred_class = (pred > 0.5).astype(int)
print(classification_report(test['Diagnosis'],pred_class))

Optimization terminated successfully.
         Current function value: 0.681340
         Iterations 4
              precision    recall  f1-score   support

           0       0.54      0.98      0.69       403
           1       0.47      0.02      0.04       347

    accuracy                           0.54       750
   macro avg       0.50      0.50      0.37       750
weighted avg       0.51      0.54      0.39       750



### 흡연 유/무

In [39]:
if model.pvalues['Smoking'] >= 0.05:
    print('p-value : ',model.pvalues['Smoking'])
    print('귀무가설, 종속 변수에 영향을 미치지 않음')
else:
    print('회귀 계수 : ',model.params['Smoking'])
    print('p-value : ',model.pvalues['Smoking'])
    print('대립가설, 종속 변수에 영향을 미침')

회귀 계수 :  -0.3367645514103481
p-value :  0.0022156857018666275
대립가설, 종속 변수에 영향을 미침


### 알코올 소비량

In [40]:
if model.pvalues['AlcoholConsumption'] >= 0.05:
    print('p-value : ',model.pvalues['AlcoholConsumption'])
    print('귀무가설, 종속 변수에 영향을 미치지 않음')
else:
    print('회귀 계수 : ',model.params['AlcoholConsumption'])
    print('p-value : ',model.pvalues['AlcoholConsumption'])
    print('대립가설, 종속 변수에 영향을 미침')

p-value :  0.3718689164862118
귀무가설, 종속 변수에 영향을 미치지 않음


### 신체 활동

In [41]:
if model.pvalues['PhysicalActivity'] >= 0.05:
    print('p-value : ',model.pvalues['PhysicalActivity'])
    print('귀무가설, 종속 변수에 영향을 미치지 않음')
else:
    print('회귀 계수 : ',model.params['PhysicalActivity'])
    print('p-value : ',model.pvalues['PhysicalActivity'])
    print('대립가설, 종속 변수에 영향을 미침')

p-value :  0.9421127934367687
귀무가설, 종속 변수에 영향을 미치지 않음


### 식단의 질

In [42]:
if model.pvalues['DietQuality'] >= 0.05:
    print('p-value : ',model.pvalues['DietQuality'])
    print('귀무가설, 종속 변수에 영향을 미치지 않음')
else:
    print('회귀 계수 : ',model.params['DietQuality'])
    print('p-value : ',model.pvalues['DietQuality'])
    print('대립가설, 종속 변수에 영향을 미침')

p-value :  0.6877973514250602
귀무가설, 종속 변수에 영향을 미치지 않음


### 수면의 질

In [43]:
if model.pvalues['SleepQuality'] >= 0.05:
    print('p-value : ',model.pvalues['SleepQuality'])
    print('귀무가설, 종속 변수에 영향을 미치지 않음')
else:
    print('회귀 계수 : ',model.params['SleepQuality'])
    print('p-value : ',model.pvalues['SleepQuality'])
    print('대립가설, 종속 변수에 영향을 미침')

p-value :  0.09830124747574867
귀무가설, 종속 변수에 영향을 미치지 않음


## 6. 인적정보 변수들이 알츠하이머의 유/무에 영향을 미친다.
- 방법 : 인적정보 변수들로 알츠하이머의 유/무를 예측하는 로지스틱 회귀 모형을 만들어 각 독립변수들의 p-value를 통해 종속 변수에 영향을 미치는지 확인
- 결과 : 로지스틱 회귀 모형의 정확성이 0.64로 높은 성능을 보이지 않고, 각 변수들 역시 종속 변수에 영향을 미치지 않음을 p-value를 통해 확인


In [44]:
from statsmodels.formula.api import logit 
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
df_lifestyle = df[['Age','Gender','Ethnicity','EducationLevel','Diagnosis']]
train,test = train_test_split(df_lifestyle,test_size = 0.3, random_state = 23)
model = logit('Diagnosis ~ Age + Gender + Ethnicity + EducationLevel',train).fit()
pred = model.predict(test)
pred_class = (pred > 0.5).astype(int)
print(accuracy_score(test['Diagnosis'],pred_class))
print(classification_report(test['Diagnosis'],pred_class))


Optimization terminated successfully.
         Current function value: 0.676601
         Iterations 4
0.5253333333333333
              precision    recall  f1-score   support

           0       0.54      0.88      0.66       403
           1       0.45      0.12      0.19       347

    accuracy                           0.53       750
   macro avg       0.49      0.50      0.43       750
weighted avg       0.50      0.53      0.44       750



### 나이

In [45]:
if model.pvalues['Age'] >= 0.05:
    print('p-value : ',model.pvalues['Age'])
    print('귀무가설, 종속 변수에 영향을 미치지 않음')
else:
    print('회귀 계수 : ',model.params['Age'])
    print('p-value : ',model.pvalues['Age'])
    print('대립가설, 종속 변수에 영향을 미침')

p-value :  0.6983490791215006
귀무가설, 종속 변수에 영향을 미치지 않음


### 성별

In [46]:
if model.pvalues['Gender'] >= 0.05:
    print('p-value : ',model.pvalues['Gender'])
    print('귀무가설, 종속 변수에 영향을 미치지 않음')
else:
    print('회귀 계수 : ',model.params['Gender'])
    print('p-value : ',model.pvalues['Gender'])
    print('대립가설, 종속 변수에 영향을 미침')

p-value :  0.44188832289364277
귀무가설, 종속 변수에 영향을 미치지 않음


### 인종

In [47]:
if model.pvalues['Ethnicity'] >= 0.05:
    print('p-value : ',model.pvalues['Ethnicity'])
    print('귀무가설, 종속 변수에 영향을 미치지 않음')
else:
    print('회귀 계수 : ',model.params['Ethnicity'])
    print('p-value : ',model.pvalues['Ethnicity'])
    print('대립가설, 종속 변수에 영향을 미침')

회귀 계수 :  -0.22778321766651505
p-value :  1.4224299793211558e-05
대립가설, 종속 변수에 영향을 미침


### 교육 수준

In [48]:
if model.pvalues['EducationLevel'] >= 0.05:
    print('p-value : ',model.pvalues['EducationLevel'])
    print('귀무가설, 종속 변수에 영향을 미치지 않음')
else:
    print('회귀 계수 : ',model.params['EducationLevel'])
    print('p-value : ',model.pvalues['EducationLevel'])
    print('대립가설, 종속 변수에 영향을 미침')

회귀 계수 :  -0.15726409844651146
p-value :  0.004269041065809874
대립가설, 종속 변수에 영향을 미침


# 결론 
- 샘플링을 통해 데이터 불균형을 줄여 f1-score가 전에 비해 상승한 것을 확인할 수 있음
- 변수들을 그룹으로 묶어 유의미성을 확인하는 것은 상호작용 효과를 간과할 수 있다는 점에서 한계가 있음
- 전체 변수들을 가지고 모델을 만든 후, 불필요한 변수들을 단계적으로 제거해나가면서 모델 성능과 변수 중요도를 확인하는 것이 더 효과적일 것이라 판단.